File: KwangHo/순위 검색.py
from itertools import combinations

def solution(info, query):
    answer = []
    
    arr={}
    #정보형태만들기
    for i in info :
        temp=i.split()
        for j in range(1,5):
            key=list(combinations(temp[:4],j))
            for k in key:
                t=''.join(k)
                if arr.get(t):
                    arr[t].append(int(temp[4]))
                else :
                    arr[t]=[]
                    arr[t].append(int(temp[4]))
        if arr.get(''):
            arr[''].append(int(temp[4]))
        else:
            arr['']=[]
            arr[''].append(int(temp[4]))
    
    #정렬
    for i in arr.values():
        i.sort()
    
    #검색
    for i in query:
        key = i.replace(' and ','')
        key = key.replace('-','')
        key = key.split(' ')
        key[1]=int(key[1])
        cnt = 0
        
        if arr.get(key[0]):
            data = arr[key[0]]
            start,end=0,len(data)
            while start!=end and start!=len(data):
                if data[(start+end)//2]>=key[1]:
                    end=(start+end)//2
                else:
                    start=(start+end)//2+1
            answer.append(len(data)-start)
        else :
            answer.append(0)
    return answer

'''
처음에는 경우수마다 돌면서 했는데 효율성에 막힘
딕셔너리로 값추가 해서 검색했는데도 효율성에서 막힘
풀이보니깐 이분검색부분이 효율성에 부딛힌점을 알고 코딩!!
(진짜 생각지도 못한 부분이였음, 데이터가 너누 많으니
이걸 첨부터 끝까지 모든 검사를 실시하니 막히는 부분이였음!)
'''

File: Doodream/순위 검색.js
function solution(info, query) {
    const answer = [];
    const infoMap = {};

    function combination(array, score, start) {
        const key = array.join("");
        const value = infoMap[key];

        if (value) {
            infoMap[key].push(score);
        } else {
            infoMap[key] = [score];
        }

        for (let i = start; i < array.length; i++) {
            const temp = [...array];
            temp[i] = "-";
            combination(temp, score, i + 1);
        }
    }

    for (const e of info) {
        const splited = e.split(" ");
        const score = Number(splited.pop());
        combination(splited, score, 0);
    }

    for (const key in infoMap) {
        infoMap[key] = infoMap[key].sort((a, b) => a - b);
    }

    for (const e of query) {
        const splited = e.replace(/ and /g, " ").split(" ");
        const score = Number(splited.pop());
        const key = splited.join("");
        const array = infoMap[key];

        if (array) {
            let start = 0;
            let end = array.length;
            while (start < end) {
                const mid = Math.floor((start + end) / 2);

                if (array[mid] >= score) {
                    end = mid;
                } else if (array[mid] < score) {
                    start = mid + 1;
                }
            }

            const result = array.length - start;
            answer.push(result);
        } else {
            answer.push(0);
        }
    }

    return answer;
}